### ISSUE TITLE
Typings

### ISSUE THREAD

**Issue by ivansieder** [auth0/node-jsonwebtoken#520]
Hi there,

first of all: THANKS A LOT for this lovely library! Really appreciate it!

I'm gonna post it here, even if it's more related to DefinitelyTyped, as it's mostly a question of understanding.

In the typings, the synchronous/asynchronous verify function seem to be able to return either an object or a string (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jsonwebtoken/index.d.ts#L126).

Now my question is, in which case can the returned payload be a string? I wasn't able to find a variation in the code, in which the returned payload would be a string. Did I miss anything there?

**Comment by MitMaro** [comment]
**tl;dr; the TypeScript typings are correct for this library, but not for JWTs in general.**

### Long explanation

This is a complicated question to answer. The JWT verify/decode functions if [strictly following the spec](https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-32#section-7.2) should always return a JWT Claim Set (object).

> Verify that the resulting octet sequence is a UTF-8 encoded
   representation of a completely valid JSON object conforming to
   RFC 7159 [RFC7159]; let the JWT Claims Set be this JSON object.

This library, while called `jsonwebtoken` technically can support any [JSON Web Signature (JWS)](https://tools.ietf.org/html/draft-ietf-jose-json-web-signature-41)`*` as it is a wrapper over [node-jws](https://github.com/brianloveswords/node-jws).

It is possible that a JWS could contain a string payload instead of an object. This library can also be configured to not verify claims, essentially skipping the JWT Claim Set validation and return the raw string. From the [JWS spec](https://tools.ietf.org/html/draft-ietf-jose-json-web-signature-41#section-2):

> JWS Payload
   The sequence of octets to be secured -- a.k.a., the message.  The
   payload can contain an arbitrary sequence of octets.

Note that, `node-jws` does [require that the payload be a JSON string ](https://github.com/brianloveswords/node-jws/blob/61f2bb9b883343c89fcb28b7ef4ed1d692110ee4/lib/verify-stream.js#L70) when `header.typ` is set to `JWT`.

From the [examples in the spec](https://tools.ietf.org/html/draft-ietf-jose-json-web-signature-41#section-3.3) for a JWS you also see the note:

>   The UTF-8 representation of following JSON object is used as the JWS Payload.  (Note that the payload can be any content, and need not be a representation of a JSON object.)

With this library, you can create a token with a non-object payload by passing a string or Buffer to the `sign` function and it will create a token with [`header.typ` not set to `JWT`](https://github.com/auth0/node-jsonwebtoken/blob/d01cc7bcbdeb606d997a580f967b3169fcc622ba/sign.js#L89). Though I think the docs are kinda misleading:

> `payload` could be an object literal, buffer or string representing valid JSON.

`*` I haven't verified this to be true, so there may be cases where this library does not support something from the JWS spec. For example "X.509 Certificate Chain" is not currently supported.

**Comment by ivansieder** [comment]
Well, that is a badass awesome explanation @MitMaro, thanks for your time and clarification! Now it's clear to me, on why it CAN return a string and a little bit of the JWT spec!
